src/utils/files.js

```diff
@@ -134,49 +50,100 @@ function treeIdGenerator(projectToImport) {
   ];
 }
 
-async function getFileFormat(fileEntry) {
-  return new Promise((resolve, reject) => fileEntry.file(resolve, reject));
+async function getFileFromFileEntry(entry) {
+  return new Promise((resolve, reject) => entry.file(resolve, reject));
 }
 
-async function handleDragAndDropFile(dirEntry) {
-  let fileList = [];
-
-  async function traverseFileTree(fileEntry) {
-    if (fileEntry.isFile) {
-      // Get file
-      return fileList.push(fileEntry);
-    } else if (fileEntry.isDirectory) {
-      // Get folder contents
-      const dirReader = fileEntry.createReader();
-      const filesEntry = await new Promise((resolve, reject) =>
-        dirReader.readEntries(resolve, reject)
-      );
-      await async.map(filesEntry, traverseFileTree);
+async function getEntriesFromDirEntry(entry) {
+  return new Promise((resolve, reject) =>
+    entry.createReader().readEntries(resolve, reject)
+  );
+}
+
+async function parseDirEntry(entry, tree = {}, files = []) {
+  if (entry.isDirectory) {
+    const childEntries = await getEntriesFromDirEntry(entry);
+    const children = await Promise.all(
+      childEntries.map(
+        async e => await parseDirEntry(e, {}, files).then(res => res.tree)
+      )
+    );
+    Object.assign(tree, {
+      name: entry.name,
+      children: children.filter(child => child.name)
+    });
+  } else {
+    files.push(await getFileFromFileEntry(entry));
+  }
+  return { tree, files };
+}
+
+async function getFilesFromEvent(event) {
+  let files, folders;
+
+  const fileEntries = [];
+  const dirEntries = [];
+
+  if (event.dataTransfer) {
+    // Files from drag & drop
+    Array.from(event.dataTransfer.items)
```

Comment:
Second argument of `Array.from` is a map function... it may be more performant to use it instead of a direct `.map` call after it.

Reply:
A simple [jsbench](https://jsben.ch/G0ha3) seems to validate what I said :) (chrome + firefox)